src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -275,164 +495,204 @@ protected override void BeginProcessing()
                 args: emptyArray);
         }
 
-        /// <summary>
-        /// Execute the processing script blocks on the current pipeline object
-        /// which is passed as it's only parameter.
-        /// </summary>
-        /// <exception cref="ParseException">Could not parse script.</exception>
-        /// <exception cref="RuntimeException">See Pipeline.Invoke.</exception>
-        /// <exception cref="ParameterBindingException">See Pipeline.Invoke.</exception>
-        protected override void ProcessRecord()
+        private void ProcessPropertyAndMethodParameterSet()
         {
-            Dbg.Assert(ParameterSetName == "ScriptBlockSet" || ParameterSetName == "PropertyAndMethodSet", "ParameterSetName is neither 'ScriptBlockSet' nor 'PropertyAndMethodSet'");
+            _targetString = string.Format(CultureInfo.InvariantCulture, InternalCommandStrings.ForEachObjectTarget, GetStringRepresentation(InputObject));
 
-            switch (ParameterSetName)
+            if (LanguagePrimitives.IsNull(InputObject))
             {
-                case "ScriptBlockSet":
-                    for (int i = _start; i < _end; i++)
+                if (_arguments != null && _arguments.Length > 0)
+                {
+                    WriteError(GenerateNameParameterError("InputObject", ParserStrings.InvokeMethodOnNull,
+                                                          "InvokeMethodOnNull", _inputObject));
+                }
+                else
+                {
+                    // should process
+                    string propertyAction = string.Format(CultureInfo.InvariantCulture,
+                        InternalCommandStrings.ForEachObjectPropertyAction, _propertyOrMethodName);
+
+                    if (ShouldProcess(_targetString, propertyAction))
                     {
-                        // Only execute scripts that aren't null. This isn't treated as an error
-                        // because it allows you to parameterize a command - for example you might allow
-                        // for actions before and after the main processing script. They could be null
-                        // by default and therefore ignored then filled in later...
-                        if (_scripts[i] != null)
+                        if (Context.IsStrictVersion(2))
```

Comment:
What does this part here mean? What does the version refer to and why is the magic number `2` used? Ideally this would self-document with names of property, method and constant, but otherwise a comment here would be desirable I think

Reply:
Old code, out of scope for this PR.